conan/tools/files/patches.py

```diff
@@ -97,16 +94,20 @@ def apply_conandata_patches(conanfile):
         - patch_file: "patches/0001-buildflatbuffers-cmake.patch"
     ```
     """
+    if conanfile.conan_data is None:
+        raise ConanException("conandata.yml not defined")
 
     patches = conanfile.conan_data.get('patches')
+    if patches is None:
+        raise ConanException("No patches defined in conandata")
```

Comment:
Why raising now? A recipe may want to always call `export_conandata_patches`, then `apply_conandata_patches`, regardless of conandata content, so that the recipe is generic and already ready for patches.

Reply:
> A recipe may want to always call export_conandata_patches, then apply_conandata_patches, regardless of conandata content, so that the recipe is generic and ready to new contributions adding patches.

I don't agree with this, this would end with all recipes in ConanCenter adding these lines, just in case it is necessary to apply a patch in the future? That adds unnecessary clutter to recipes. If we are not adding these lines to all recipes, then it will only be added to recipes that have patches, and accordingly have a ``patches`` field defined in the ``conandata.yml``. I think using a method ``export_conandata_patches()`` in a recipe implies a pre-condition that there are ``patches`` in ``conandata`` to export, otherwise, why are you calling that method?

In some exceptional case, if you want to be future-ready, you can still define the ``patches: {}`` in the conadata.yml

Reply:
But it raises now in `apply_conandata_patches` as well, and as I said the reducer removes `patches` field when there is no patch for a specific version (there might be patches for a version and not for others, this is why we fallback to empty dic/list in each `get` in the loop we do in CCI recipes).

Reply:
> In some exceptional case, if you want to be future-ready, you can still define the patches: {} in the conadata.yml

The advantage to be future-ready is to avoid to modify `conanfile.py`, and thus RREV of versions not affected by a new patch file.

Reply:
But the code is only requesting that there is a field ``patches`` in the ``conandata.yml``. Not that it is filled with patches. It is fine if some version doesn't have patches. But if I am using a ``xxx_conandata_patches``, the expectation is that the ``conandata.yml`` will contain a ``patches`` section, and not a ``patchs`` or a ``patch`` or a ``pathces`` one (or no one).

Reply:
There is no `patches` field in the `conandata.yml` in local cache after a `conan create`/`conan export` if the specific exported version has no patch.

Please test with cjson in CCI for example.
See the conandata.yml: https://github.com/conan-io/conan-center-index/blob/624d6cfa702eecd4f0af9d32c1d33e4e409f7631/recipes/cjson/all/conandata.yml
There is a `patches` field, but only a patch for 1.7.12. Try to export 1.7.15, check conandata.yml in local cache: no `patches` field.

Reply:
Ok, I see the problem. The ConanCenter "reducer" hook, is completely dropping the keys when the version is not defined. This is unfortunate, I didn't expect this, I think the "reducer" should reduce things to the current version information, but not completely drop a full section. 

I'll relax that assumption.

Reply:
Changed, now it will only print one informative message telling that there are no patches to apply (that might help to debug some cases)

Reply:
Looking good now @SpaceIm?

Reply:
yes